lib/srv/regular/sshserver.go

```diff
@@ -1096,40 +1109,254 @@ func (s *Server) getServerResource() (types.Resource, error) {
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// The forwarding process may have chosen its own port, so we need to get the
+	// new listen address.
+	var fn int
+	fbuf := make([]*os.File, 1)
+	readDone := make(chan struct{})
+	// Read addr in another goroutine so we can cancel it if the forwarding process
+	// stops.
+	go func() {
+		_, fn, err = localConn.ReadWithFDs(nil, fbuf)
+		close(readDone)
+	}()
+
+	select {
+	case <-proc.Done:
+		localConn.Close()
+		return nil, trace.Errorf("forwarding process is closed")
+	case <-readDone:
+	}
+
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if fn == 0 {
+		return nil, trace.BadParameter("forwarding process did not return a listener")
+	}
+	if err := validateListenerSocket(scx, localConn.UnixConn, fbuf[0]); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := net.FileListener(fbuf[0])
+	return listener, trace.Wrap(err)
+}
+
+func controlSyscallConn(conn syscall.RawConn, f func(fd uintptr) error) error {
+	var err error
+	err2 := conn.Control(func(fd uintptr) {
+		err = f(fd)
+	})
+	return trace.NewAggregate(err2, err)
+}
+
+// validateListenerSocket checks that the socket and listener file descriptor
+// sent from the forwarding process have the expected properties.
+func validateListenerSocket(scx *srv.ServerContext, controlConn *net.UnixConn, listenerFD *os.File) error {
+	// Get the credentials of the client connected to the socket.
+	syscallConn, err := controlConn.SyscallConn()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	var cred *unix.Ucred
+	if err := controlSyscallConn(syscallConn, func(fd uintptr) error {
+		cred, err = unix.GetsockoptUcred(int(fd), unix.SOL_SOCKET, unix.SO_PEERCRED)
+		return err
+	}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Check that the user connected to the socket is who we expect.
+	usr, err := user.Lookup(scx.Identity.Login)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if expectedUid, err := strconv.Atoi(usr.Uid); err != nil {
+		return trace.Wrap(err)
+	} else if int(cred.Uid) != expectedUid {
+		return trace.AccessDenied("unexpected user UID for the socket: %v", cred.Uid)
+	}
+	if expectedGid, err := strconv.Atoi(usr.Gid); err != nil {
+		return trace.Wrap(err)
+	} else if int(cred.Gid) != expectedGid {
+		return trace.AccessDenied("unexpected user GID for the socket: %v", cred.Gid)
+	}
+
+	// Verify that the listener is a socket.
+	if fileInfo, err := listenerFD.Stat(); err != nil {
+		return trace.Wrap(err)
+	} else if fileInfo.Mode()&os.ModeSocket == 0 {
+		return trace.AccessDenied("file descriptor is not a socket")
+	}
+
+	// Verify the socket type
+	if sockType, err := unix.GetsockoptInt(int(listenerFD.Fd()), unix.SOL_SOCKET, unix.SO_TYPE); err != nil {
+		return trace.Wrap(err)
+	} else if sockType != unix.SOCK_STREAM && sockType != unix.SOCK_SEQPACKET {
+		return trace.AccessDenied("socket is not of the expected type (STREAM or SEQPACKET)")
+	}
```

Comment:
Why would we want a SEQPACKET socket?

```suggestion
	} else if sockType != unix.SOCK_STREAM {
		return trace.AccessDenied("expected socket type SOCK_STREAM, got %x", sockType)
	}
```